ebpf/state.go

```
@@ -0,0 +1,209 @@
+		conns[i].LastRecvBytes = prev.lastRecv
+		conns[i].LastRetransmits = prev.lastRetransmits
+	}
+
+	return conns
+}
+
+func (ns *networkState) StoreConnections(conns []ConnectionStats) {
+	// Update connections
+	ns.connsMutex.Lock()
+	defer ns.connsMutex.Unlock()
+	ns.connections = conns
+}
+
+// StoreClosedConnection stores the given connection for every client
+func (ns *networkState) StoreClosedConnection(conn ConnectionStats) {
+	ns.clientsMutex.Lock()
+	defer ns.clientsMutex.Unlock()
+
+	for id := range ns.clients {
+		// We only store the pointer to the connection, when it will be cleared for each client it will get GCed
+		ns.clients[id].closedConnections = append(ns.clients[id].closedConnections, &conn)
+	}
+}
+
+// closedConns returns the closed connections for the given client and takes care of updating last fetch
+// the provided client is supposed to exist
+func (ns *networkState) closedConns(clientID int) []ConnectionStats {
+	conns := []ConnectionStats{}
+
+	ns.clientsMutex.Lock()
+	defer ns.clientsMutex.Unlock()
+
+	for _, conn := range ns.clients[clientID].closedConnections {
+		conns = append(conns, *conn)
+	}
+
+	// Flush closed connections for this client
+	ns.clients[clientID].closedConnections = []*ConnectionStats{}
+	ns.clients[clientID].lastFetch = time.Now()
+	return conns
+}
+
+// newClient creates a new client and returns true if the given client already exists
+func (ns *networkState) newClient(clientID int) bool {
+	ns.clientsMutex.Lock()
+	defer ns.clientsMutex.Unlock()
+	if _, ok := ns.clients[clientID]; ok {
+		return true
+	}
+
+	ns.clients[clientID] = &client{
+		id:        clientID,
+		lastFetch: time.Now(),
+		stats:     map[string]*sendRecvStats{},
+	}
+	return false
+}
+
+func (ns *networkState) RemoveClient(clientID int) error {
+	ns.clientsMutex.Lock()
+	defer ns.clientsMutex.Unlock()
+
+	if _, ok := ns.clients[clientID]; !ok {
+		return fmt.Errorf("can't remove client %d, it is not stored", clientID)
```

Thoughts on failing silently here?